File: Easy/045_MajorElement/C++/main.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <map>

void majorElement (std::string &);

int main (int argc, char **argv)
{
    // if argc is less than 2, no file passed
    if (argc < 2)
        // output message
        printf("Usage: ./a.out file.txt\n");

    // grab file user passed to program
    std::fstream file (argv[1], std::ios::in);

    // if the file is good
    if (file)
    {
        // declare a string to hold each line
        std::string line;

        // while we can get a line from the file
        while (std::getline(file, line))
        {
            // call method on current line
            majorElement(line);
        }
    }

    return 0;
}

/*
    This method takes the line and finds the major element
*/
void majorElement(std::string &line)
{

    // create a map to hold the numbers and the count
    std::map<std::string, int> map;
    // init a string stream on the line
    std::stringstream ss(line);
    // declare a token
    std::string token;
    // init a count to 0
    int count = 0;
    // while we can parse the line on a ','
    while(std::getline(ss, token, ','))
    {
        // initialize a map iterator
        std::map<std::string, int>::iterator it;
        // find the token
        it = map.find(token);
        // if found
        if (it != map.end())
        {
            // get the current count
            int count = map[token];
            // update count
            map[token] = ++count;
        }
        else
        {
            // else, set count to 1
            map[token] = 1;
        }
        // increment the count
        count++;
    }
    
    // declare a string to hold the num
    std::string num;
    // declare a boolean to false, we haven't found the num
    bool found = false;
    // init a iterator at the begining of the map
    std::map<std::string, int>::iterator it = map.begin();
    // loop through the map
    for (std::pair<std::string, int>ele : map)
    {
        // if the count is atleast 1/2 the length of numbers
        if (ele.second >= (count/2))
        {
            // set the num
            num = ele.first;
            // we found one
            found = true;
        }
    }
    // if found, output the num, else output none to stdout
    if (found)
        std::cout << num << std::endl;
    else
        std::cout << "None" << std::endl;

}

File: Easy/045_MajorElement/Java/Main.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.*;

public class Main
{

    public static void main(String[] args) throws IOException
    {

        // create a list to hold each line of the file
        ArrayList<String> list = new ArrayList<>();

        if (args.length > 0)
        {
            // create a reader to read the file
            BufferedReader reader = new BufferedReader(new FileReader(args[0]));

            // init a string to hold each line
            String line = "";

            // while able to get a line from the file, add to the list
            while ((line = reader.readLine()) != null)
                list.add(line);

            // close out the reader
            reader.close();

            // for each item in the list
            for (String item : list)
            {
                majorElement(item);
            }
        }
    }

    /*
        This method prints to stdout the major element
        of a set of numbers, where the set looks like:
        2,3,3,4,5,5,9,1,89,9,3

        If a number has count that is greater than half
        the number of elements, then its the major element

     */
    private static void majorElement(String line)
    {
        // split the line on commas, into an array of strings
        String data[] = line.split(",");

        // init a new Hash Map that will contain the number
        // and the count of that number
        HashMap<String, Integer> countMap = new HashMap<>();

        // init a string to hold what the major element is
        String major = "";
        // init a boolean to keep track if we found a major element
        boolean hasMajor = false;

        // loop over the array of numbers
        for (String num : data)
        {

            // if we have this key
            if (countMap.containsKey(num))
            {
                // increment the count
                int count = countMap.get(num);
                countMap.put(num, ++count);

                // if this count is a major element
                if (++count >= data.length/2)
                {
                    // set the bool, and string; then break
                    hasMajor = true;
                    major = num;
                    break;
                }
            }
            // otherwise, put count of the current number : 1
            else
                countMap.put(num, 1);
        }

        // if we have a major, then output the number
        if (hasMajor)
            System.out.println(major);
        // otherwise output None
        else
            System.out.println("None");


    }
}
